persistence/remote/src/main/java/org/infinispan/persistence/remote/RemoteStore.java

```diff
@@ -68,228 +72,257 @@
+                  if (localValueStorageType.isBinary()) {
+                     dataFormatBuilder.valueType(localValueStorageType);
+                     dataFormatBuilder.valueMarshaller(IdentityMarshaller.INSTANCE);
+                  } else {
+                     dataFormatBuilder.valueType(marshaller.mediaType());
+                     dataFormatBuilder.valueMarshaller(marshaller);
+                  }
+               }
+               DataFormat dataFormat = dataFormatBuilder.build();
+               log.fatalf("Data format is %s", dataFormat);
+               remoteCache = remoteCache.withDataFormat(dataFormat);
+            });
    }
 
    @Override
-   public void stop() throws PersistenceException {
-      remoteCacheManager.stop();
+   public Set<Characteristic> characteristics() {
+      Set<Characteristic> characteristics = EnumSet.of(Characteristic.BULK_READ, Characteristic.EXPIRATION,
+            Characteristic.SHAREABLE);
+      if (supportsSegmentation) {
+         characteristics.add(Characteristic.SEGMENTABLE);
+      }
+      return characteristics;
    }
 
    @Override
-   public boolean isAvailable() {
-      try {
-         PingResponse pr = ((RemoteCacheImpl) remoteCache).ping();
-         return pr.isSuccess();
-      } catch (Exception e) {
-         return false;
-      }
+   public CompletionStage<Void> stop() {
+      return blockingManager.runBlocking(() -> remoteCacheManager.stop(), "RemoteStore-stop");
    }
 
    @Override
-   public MarshallableEntry<K, V> get(int segment, Object key) {
-      return loadEntry(key);
+   public CompletionStage<Boolean> isAvailable() {
+      return remoteCache.ping()
+            .handle((v, t) -> t == null && v.isSuccess());
    }
 
    @Override
-   public MarshallableEntry<K, V> loadEntry(Object key) throws PersistenceException {
+   public CompletionStage<MarshallableEntry<K, V>> load(int segment, Object key) {
       if (configuration.rawValues()) {
-         Object unwrappedKey;
-         if (key instanceof WrappedByteArray) {
-            unwrappedKey = ((WrappedByteArray) key).getBytes();
-         } else {
-            unwrappedKey = key;
-         }
-         MetadataValue<?> value = remoteCache.getWithMetadata(unwrappedKey);
-         if (value != null) {
-            Metadata metadata = new EmbeddedMetadata.Builder()
-                  .version(new NumericVersion(value.getVersion()))
-                  .lifespan(value.getLifespan(), TimeUnit.SECONDS)
-                  .maxIdle(value.getMaxIdle(), TimeUnit.SECONDS).build();
-            long created = value.getCreated();
-            long lastUsed = value.getLastUsed();
-            Object realValue = value.getValue();
-            if (realValue instanceof byte[]) {
-               realValue = new WrappedByteArray((byte[]) realValue);
+         Object unwrappedKey = unwrap(key);
+         CompletableFuture<MetadataValue<Object>> valueStage = remoteCache.getWithMetadataAsync(unwrappedKey);
```

Comment:
Nitpick, but unnecessary variable as the declaration is longer than the actual method call. I would just call `thenApply` directly.

Reply:
Sure.